### PR TITLE
[python-package] do not copy column-major numpy arrays when creating Dataset

### PR DESCRIPTION
**Pull request by jmoralez**
Changes the logic to flatten the data array to consider its layout:

* If the array is Fortran contiguous, then it's flattened as such and passed to the Dataset constructor indicating that the array is column-major, which avoids copying it to make it row-major.
* If the array is C contiguous, no copy is made (current behavior).
* If the array is not contiguous because it was sliced, then a copy is made to make it C-contiguous (current behavior).

I ran some quick tests with an array of (100k, 50) and the timings to construct the dataset from a C-contiguous and an F-contiguous arrays are roughly the same, so this doesn't introduce extra latency, just avoids the copy.

### CODE REVIEW

**Comment by jmoralez** [comment]
This could also be done for the predict portion, i.e.

https://github.com/microsoft/LightGBM/blob/5151fe85f08e5dccff7d48242dddace51f9c8ede/python-package/lightgbm/basic.py#L1268-L1307

Please let me know if it's ok to include it here or in a separate PR.

File: tests/python_package_test/test_basic.py [hunk 1]
```diff
@@ -947,3 +947,16 @@ def test_max_depth_warning_is_raised_if_max_depth_gte_5_and_num_leaves_omitted(c
         "in params. Alternatively, pass (max_depth=-1) and just use 'num_leaves' to constrain model complexity."
     )
     assert expected_warning in capsys.readouterr().out
+
+
+@pytest.mark.parametrize("order", ["C", "F"])
+@pytest.mark.parametrize("dtype", ["float32", "int64"])
+def test_no_copy_in_dataset_from_numpy_2d(order, dtype):
+    X = np.random.default_rng().random(size=(100, 3))
```
Comment (StrikerRUS) [inline tests/python_package_test/test_basic.py:955]: Maybe reuse `rng` fixture?
https://github.com/microsoft/LightGBM/blob/83c0ff3de1925b0e2d4831a9ccb6ffc196aa795b/tests/python_package_test/conftest.py#L5-L7

**Review by StrikerRUS — APPROVED** [review]
LGTM! Thank you for very useful improvement!

I just left a couple of non-blocking comments below for your consideration.

**Comment by StrikerRUS** [comment]
> Please let me know if it's ok to include it here or in a separate PR.

I believe it would be better to have a follow-up PR for prediction after merging this one.

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -4611,3 +4611,26 @@ def test_bagging_by_query_in_lambdarank():
...
+    for tree in dumped_row["tree_info"]:
+        assert tree["num_leaves"] > 1
+
+    X_col = np.asfortranarray(X_row)
+    assert X_col.flags["F_CONTIGUOUS"]
+    ds_col = lgb.Dataset(X_col, y)
+    bst_col = lgb.train(params, ds_col, num_boost_round=rounds)
+    dumped_col = bst_col.dump_model()
+
+    assert dumped_row == dumped_col
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:4636]: Maybe also check that `ds_row` and `ds_col` are identical?
https://github.com/microsoft/LightGBM/blob/83c0ff3de1925b0e2d4831a9ccb6ffc196aa795b/python-package/lightgbm/basic.py#L3535-L3536

**Review by jameslamb — COMMENTED** [review]
This looks great to me! I agree with all of @StrikerRUS 's comments, so won't approve until those are addressed. But I don't have any additional comments.

Thanks for working on this, and really nice tests!

File: tests/python_package_test/test_engine.py [hunk 3]
```diff
@@ -4611,3 +4612,22 @@ def test_bagging_by_query_in_lambdarank():
     ndcg_score_no_bagging_by_query = gbm_no_bagging_by_query.best_score["valid_0"]["ndcg@5"]
     assert ndcg_score_bagging_by_query >= ndcg_score - 0.1
     assert ndcg_score_no_bagging_by_query >= ndcg_score - 0.1
+
+
+def test_equal_datasets_from_row_major_and_col_major_data(tmp_path):
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:4617]: As this test now doesn't include `train()`, I think it should go to `test_basic.py` file.

Comment (jmoralez) [reply tests/python_package_test/test_engine.py]: Sorry. Moved in 38c6786

**Comment by StrikerRUS** [comment]
@jameslamb 
>I agree with all of @StrikerRUS 's comments, so won't approve until those are addressed.

Would you like to take a look after the recent commits?

**Review by jameslamb — APPROVED** [review]
Changes look great to me, thank you!

I looked at the failing Dask tests... they're unrelated to this PR. Documented in #6739.

@jmoralez do you have time in the next few days to investigate that?

**Comment by jmoralez** [comment]
Sure, I'll take a look.

**Pull request merged**